### PR TITLE
Implement ModuleSearchBase for enhanced DNN Platform Search

### LINKED ISSUE #94: Integrate DNN Search
## Is your feature request related to a problem?
The Forum Content does not get indexed by DNN Search

## Describe the solution you'd like
Integrate the (once new) DNN Search.
See also: https://github.com/ActiveForums/ActiveForums/issues/12

### PR DESCRIPTION
**Pull request by johnhenley**
### Implements the improved DNN Search experience for Community Forums. 
https://www.dnnsoftware.com/community-blog/cid/154913/Integrating-with-Search-Introducing-ModuleSearchBase
https://www.dnnsoftware.com/community-blog/cid/153947/search-done-right-in-dnn

## Changes made
- Convert TopicsController to implement ModuleSearchBase rather than ISearchable to implement newer search functionality introduced in DNN Platform 7.1.
- If you are using the soft-delete behavior in Community Forums, searches are now incrementally performed based on content update date including marking deleted content as inactive in the search meta data. Otherwise, full rebuild is done at each scheduled search crawler interval.
- If you utilize tags within Community Forums, these will be included in the search metadata.
- View permissions from the forum's security settings are included in the search metadata, so only authorized forum topic viewers should see topics within search results.
- NOTE: In search results, the "source" display name comes from Module's display name, e.g. "Active Forums". If you want to override this, add an entry to ~/DesktopModules/Admin/SearchResults/App_LocalResources/SearchableModules.resx for the Module_[MODULENAME].text, e.g., <data name="Module_Active Forums.Text" xml:space="preserve"><value>Community Forums</value></data>
-  A possible future enhancement might be to write this entry or to perhaps change the module definition ...
			
## PR Template Checklist

- [ ] Fixes Bug
- [X] Feature solution
- [ ] Other

## Please mark which issue is solved

Close #94

### CODE REVIEW

**Comment by johnhenley** [comment]
@WillStrohl Try as I might, I still have some 'git cruft' for you to deal with...for some reason, this pull request pulls in changes from #121 that separated out the Topics.cs into TopicInfo.cs and TopicController.cs. Not sure why. Also, some whitespace issues, but I'm sure you can deal with those. Lastly, since I still don't have a way to separate the SQL updates per pull request, and I still have 3 or 4 of them outstanding, this one includes all of the SQL updates from this pull request as well as the other ones. Let me know if you need help sorting that out.

**Comment by WillStrohl** [comment]
FYI - I've been trying to use GitHub to attempt to resolve the conflicts so this can be merged, but GitHub crashes every time I try.  :(

**Review by WillStrohl — CHANGES_REQUESTED** [review]

File: components/Data/Common.cs [hunk 1]
```diff
@@ -84,6 +84,37 @@ public string CheckForumIdsForView(string ForumIds, string UserRoles)
 
 			return sForums;
 		}
+		public string WhichRolesCanViewForum(int ForumId, string UserRoles)
+		{
+			string cacheKey = string.Format("AF-CanView-{0}", ForumId.ToString());
+			string sSQL = "SELECT f.ForumId, ISNULL(CanView,'') as CanView from " + dbPrefix + "Permissions as P INNER JOIN " + dbPrefix + "Forums as f on f.PermissionsID = P.PermissionsId WHERE f.ForumId = " + ForumId.ToString();
```
[the remark this answers is not in the document]
Comment (johnhenley) [reply components/Data/Common.cs:90]: @WillStrohl 
I literally almost copied that method from the one above it, which also has the same embedded SQL. So I would suggest this is something that should be punted to a new issue and handled with a separate pull request?
image

**Review by WillStrohl — APPROVED** [review]
We'll accept this PR as-is, but it should be noted that we've added to our technical debt with this update.

File: components/Data/Common.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnhenley) [reply components/Data/Common.cs:90]: I will create an issue to pay back the debt :)

[the remark this answers is not in the document]
Comment (johnhenley) [reply components/Data/Common.cs:90]: >  Also, it's missing DNNs (still) required best practice of including the DNN tokens for compatibility with anyone that may be using a different schema (`{databaseOwner}`) or object qualifier (`{objectQualifier]`).

Actually, that accomplished in a round-about way by **_`" + dbPrefix + "`_** part in the concatenated SQL. The dbPrefix comes from:
image
which comes from here:
image

**Pull request merged**